fdb-extensions/src/main/java/com/apple/foundationdb/async/RTree.java

```diff
@@ -0,0 +1,1878 @@
+        if (insertSlotIndex.get() < 0) {
+            // just update the slot with the potentially new value
+            writeNodeSlot(transaction, targetNode.getId(), newSlot);
+            return AsyncUtil.DONE;
+        }
+
+        final AtomicReference<Node> currentNode = new AtomicReference<>(targetNode);
+        final AtomicReference<NodeSlot> parentSlot = new AtomicReference<>(newSlot);
+
+        return AsyncUtil.whileTrue(() -> {
+            final NodeSlot currentNewSlot = parentSlot.get();
+
+            if (currentNewSlot != null) {
+                return insertSlotIntoTargetNode(transaction, currentNode.get(), currentNewSlot, insertSlotIndex.get())
+                        .thenApply(nodeOrAdjust -> {
+                            if (currentNode.get().isRoot()) {
+                                return false;
+                            }
+                            currentNode.set(currentNode.get().getParentNode());
+                            parentSlot.set(nodeOrAdjust.getSlotInParent());
+                            insertSlotIndex.set(nodeOrAdjust.getSplitNode() == null ? -1 : nodeOrAdjust.getSplitNode().getSlotIndexInParent());
+                            return nodeOrAdjust.getSplitNode() != null || nodeOrAdjust.parentNeedsAdjustment();
+                        });
+            } else {
+                // adjustment only
+                final NodeOrAdjust splitNodeSlotOrAdjust = updateSlotsAndAdjustNode(transaction, currentNode.get());
+                Verify.verify(splitNodeSlotOrAdjust.getSlotInParent() == null);
+                if (currentNode.get().isRoot()) {
+                    return AsyncUtil.READY_FALSE;
+                }
+                currentNode.set(currentNode.get().getParentNode());
+                return splitNodeSlotOrAdjust.parentNeedsAdjustment()
+                       ? AsyncUtil.READY_TRUE
+                       : AsyncUtil.READY_FALSE;
+            }
+        }, executor);
+    }
+
+    @Nonnull
+    private CompletableFuture<NodeOrAdjust> insertSlotIntoTargetNode(@Nonnull final Transaction transaction,
+                                                                     @Nonnull final Node targetNode,
+                                                                     @Nonnull final NodeSlot newSlot,
+                                                                     final int slotIndexInTargetNode) {
+        if (targetNode.getSlots().size() < config.getMaxM()) {
+            logger.trace("regular insert without splitting; node={}; size={}", bytesToHex(targetNode.getId()), targetNode.size());
+            targetNode.insertSlot(slotIndexInTargetNode, newSlot);
+
+            if (targetNode.getKind() == Kind.INTERMEDIATE) {
+                // if this is an insert for an intermediate node, the child node referred to by newSlot
+                // is a split node from a lower level and all slots need to be updated
```

Comment:
What's the bit about "all slots need to be updated" mean here? Is there something about the way the tree is maintained that requires that all slots be updated?

Reply:
All slots of the target node need to be updates as the newly inserted node comes from a split operation meaning that the sibling nodes and the target nodes have already changed when we dealt on with them while while creating the split on a lower level. All siblings have to at least be rewritten. We currently write them all so there is definitely room fro improvement. Left a `TODO`.